File: examples/word-add-image.ts
/**
 * 例子（Epic-12）：从零构造一份 docx，在段落里嵌一张 PNG 图片，写到指定路径。
 *
 * 跑法：
 *   bun run examples/word-add-image.ts <output.docx>
 */

import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { Paragraph, WordprocessingDocument, createImageRunForWord } from "../src/word/index.js";

const HERE = dirname(fileURLToPath(import.meta.url));

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: word-add-image <output.docx>\n");
    process.exit(2);
  }

  // 用项目自带的 fixture PNG（任何 PNG 都行）。
  const pngPath = join(HERE, "..", "test", "fixtures", "tiny.png");
  let png: Uint8Array;
  try {
    png = new Uint8Array(readFileSync(pngPath));
  } catch {
    // fallback：内联一份最小 PNG（1×1 透明像素）
    png = new Uint8Array([
      0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a, 0x00, 0x00, 0x00, 0x0d, 0x49, 0x48, 0x44,
      0x52, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x08, 0x06, 0x00, 0x00, 0x00, 0x1f,
      0x15, 0xc4, 0x89, 0x00, 0x00, 0x00, 0x0d, 0x49, 0x44, 0x41, 0x54, 0x78, 0x9c, 0x62, 0x00,
      0x01, 0x00, 0x00, 0x05, 0x00, 0x01, 0x0d, 0x0a, 0x2d, 0xb4, 0x00, 0x00, 0x00, 0x00, 0x49,
      0x45, 0x4e, 0x44, 0xae, 0x42, 0x60, 0x82,
    ]);
  }

  const doc = WordprocessingDocument.create();
  const { relId } = doc.addImagePart(png);

  // 1 inch ≈ 914400 EMU；这里放 1×1 inch
  const imageRun = createImageRunForWord(relId, 914400, 914400, {
    name: "Logo",
    descr: "openxml-ts demo logo",
  });

  // 找到 Body（mainDocumentPart > Document > Body），加一段含图片的 Paragraph
  const body = doc.mainDocumentPart?.document.firstChild()!;
  const para = new Paragraph();
  para.appendChild(imageRun);
  (body as { appendChild: (e: unknown) => void }).appendChild(para);

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Wrote ${outputPath} (${png.byteLength}-byte PNG embedded)\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/WordAddImage.cs
// Replica of examples/word-add-image.ts
// Creates a docx with a paragraph containing an embedded 1x1 transparent PNG (1 inch x 1 inch).

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;
using A = DocumentFormat.OpenXml.Drawing;
using DW = DocumentFormat.OpenXml.Drawing.Wordprocessing;
using PIC = DocumentFormat.OpenXml.Drawing.Pictures;

namespace CrossSdkVerify.Replicas;

public static class WordAddImage
{
    // 1x1 transparent PNG (~68 bytes) - same as in the TS example
    private static readonly byte[] TinyPng =
    [
        0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a, 0x00, 0x00, 0x00, 0x0d, 0x49, 0x48, 0x44,
        0x52, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x08, 0x06, 0x00, 0x00, 0x00, 0x1f,
        0x15, 0xc4, 0x89, 0x00, 0x00, 0x00, 0x0d, 0x49, 0x44, 0x41, 0x54, 0x78, 0x9c, 0x62, 0x00,
        0x01, 0x00, 0x00, 0x05, 0x00, 0x01, 0x0d, 0x0a, 0x2d, 0xb4, 0x00, 0x00, 0x00, 0x00, 0x49,
        0x45, 0x4e, 0x44, 0xae, 0x42, 0x60, 0x82,
    ];

    public static void Run(string outputPath)
    {
        using var doc = WordprocessingDocument.Create(outputPath,
            DocumentFormat.OpenXml.WordprocessingDocumentType.Document);

        var mainPart = doc.AddMainDocumentPart();
        mainPart.Document = new Document();
        var body = new Body();
        mainPart.Document.Append(body);

        // Add image part
        var imagePart = mainPart.AddImagePart(ImagePartType.Png);
        using (var ms = new MemoryStream(TinyPng))
            imagePart.FeedData(ms);
        var relId = mainPart.GetIdOfPart(imagePart);

        // 1 inch = 914400 EMU
        const long emuW = 914400L;
        const long emuH = 914400L;

        var drawing = new Drawing(
            new DW.Inline(
                new DW.Extent { Cx = emuW, Cy = emuH },
                new DW.EffectExtent { LeftEdge = 0, TopEdge = 0, RightEdge = 0, BottomEdge = 0 },
                new DW.DocProperties { Id = 1, Name = "Logo", Description = "openxml-ts demo logo" },
                new DW.NonVisualGraphicFrameDrawingProperties(
                    new A.GraphicFrameLocks { NoChangeAspect = true }),
                new A.Graphic(
                    new A.GraphicData(
                        new PIC.Picture(
                            new PIC.NonVisualPictureProperties(
                                new PIC.NonVisualDrawingProperties { Id = 0, Name = "Logo" },
                                new PIC.NonVisualPictureDrawingProperties()),
                            new PIC.BlipFill(
                                new A.Blip { Embed = relId },
                                new A.Stretch(new A.FillRectangle())),
                            new PIC.ShapeProperties(
                                new A.Transform2D(
                                    new A.Offset { X = 0, Y = 0 },
                                    new A.Extents { Cx = emuW, Cy = emuH }),
                                new A.PresetGeometry(new A.AdjustValueList())
                                { Preset = A.ShapeTypeValues.Rectangle })))
                    { Uri = "http://schemas.openxmlformats.org/drawingml/2006/picture" })
            )
            { DistanceFromTop = 0, DistanceFromBottom = 0, DistanceFromLeft = 0, DistanceFromRight = 0 }
        );

        body.Append(new Paragraph(new Run(drawing)));
        body.Append(new SectionProperties());
        mainPart.Document.Save();
    }
}
